third_party/torch-mlir-dialects/README.md

```diff
@@ -0,0 +1,11 @@
+# Torch-MLIR Dialects Project
+
+Sources for torch-mlir's public dialects (containing ops/types/attributes that
+are unique to Torch-MLIR at the moment)
+
+This project is intended to be used via LLVM's external projects setup:
+
+* `-DLLVM_EXTERNAL_PROJECTS=torch-mlir-dialects`
+* `-DLLVM_EXTERNAL_TORCH_MLIR_DIALECTS_SOURCE_DIR={this_directory}`
+
+It depends on the `mlir` project.
```

Comment:
So these files are coming from https://github.com/llvm/torch-mlir? Please include that information in the README here, and also include instructions for updating these files (which directories to copy, which files to exclude).

Could we use a git submodule instead? Most of https://github.com/google/iree/tree/main/third_party is set up that way, though large optional deps that don't often require local changes like TensorFlow and Tint are fetched on demand by the build system:

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/integrations/tensorflow/WORKSPACE#L12-L16

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/build_tools/third_party/tint/CMakeLists.txt#L30-L31

How frequently will these files need to be updated? Will the files only be updated upstream then pulled here, or will local modifications be needed?

Reply:
Hey Scott, we didn't go with the submodule approach because this is a temporary solution and will all be removed once the tmtensor dialect is upstreamed to core which is expected to happen in Q2.

> How frequently will these files need to be updated? 

I don't expect the files would be updated very often if at all. Technically, this depends on what LinalgExt ops are needed to run the targeted torch models or if LinalgExt ops (scan and scatter op for now) interface changes in IREE. The workflow is a bit weird at this point: `a LinalgExt op is needed or its interface changed -> copy the op ods to torch-mlir-dialects as a tmtensor op -> copy torch-mlir-dialects to iree`. However, the scatter op is the only one we are aware of that's actually needed in the near future. 

> Will the files only be updated upstream then pulled here, or will local modifications be needed?

The files should be updated upstream and then pulled here 
- with some directories removed like `test`, `tools`, `lib/Dialect/TMTensor/Transforms`, `include/torch-mlir-dialects/Dialect/TMTensor/Transforms`. 
- with format changed according to IREE's standard. Torch-mlir uses different  `clang-format`.
- llvm version bump: torch-mlir will try best to keep up with llvm version used by IREE. However,  it should be OK to apply changes related to llvm version bumping to torch-mlir-dialects.

Reply:
> we didn't go with the submodule approach because this is a temporary solution and will all be removed once the tmtensor dialect is upstreamed to core which is expected to happen in Q2

Files exist in the project's git history forever, while submodules are pretty cheap to add and remove, though they do take time to fetch from when they exist. @GMNGeoffrey any thoughts on this?

> * with format changed according to IREE's standard. Torch-mlir uses different clang-format.

we explicitly disable formatting in third_party/, and pristine copies of files are much easier to work with, so you can remove that step

https://github.com/google/iree/blob/47e45d3c433a4b20eb9ef28c2340e47ae7108ca0/third_party/.clang-format#L15-L17

Reply:
Submodules don't _need_ to be narrow projects, but that can be nice when possible. For example, https://github.com/tensorflow/mlir-hlo is a standalone repository mirroring the [tensorflow/compiler/mlir/hlo](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/mlir/hlo) folder in tensorflow's full repo. On the other hand, we include all of [llvm/llvm-project](https://github.com/llvm/llvm-project/) as a submodule.

Another benefit to using submodules is that they can be updated in the same way (e.g. https://stackoverflow.com/a/5828396)

Reply:
I dont know if `torch-mlir` can be made a submodule of IREE. Torch-mlir uses IREE and this will add a cyclic dependence. The only way to break the dependence is to move the parts that are needed to connect to IREE into its own repository from where Torch and IREE use it. Kind of like MLIR-HLO.
Better voices might make a more reasoned decision here. When discussing with Stella, she suggesting just copying things instead of submodules. If you ask me, I will go with what Stella says, but you guys know better than I do.

Reply:
> Torch-mlir uses IREE and this will add a cyclic dependence.

This no longer appears to be the case as of https://github.com/llvm/torch-mlir/commit/1a0b953ea7edfaf26fd3881d102c64fe0c1aab29?

Cycles are also only an issue if you recursively fetch submodules, which we discourage. That is a good point for projects downstream from IREE though (they may opt for a recursive submodule fetch, which we don't want to break).

> Though not sure how this would work with LLVM bumps.

This has to be solved regardless of how the code exists in the repo. If we fork the code into IREE (as this PR does) then we'd need to maintain that fork directly. If we pull a submodule then we need to update that submodule as we do our other LLVM-dependent submodules.

Reply:
How does the llvm bump for other submodule projects in IREE work? If IREE wants to bump llvm to a specific version but a submodule project is not at that version yet would IREE requires the upstream project to bump and then pull in the changes? or someone from IREE needs to fix all the related code for the submodule within IREE?

Reply:
Yep, since LLVM and MLIR offer no compatibility guarantees. We have https://github.com/google/iree-mhlo-fork, https://github.com/google/iree-llvm-fork, and https://github.com/google/iree-tf-fork in case we need to carry forward local fixes. Each dependency increases this cost.

As it relates to this new dependency, we have a few options:

1) Fork the code into third_party. Start with a pristine copy then include local changes as part of bumping the LLVM version. These diffs will be difficult to contribute back to upstream and diverging will be costly if development continues in two locations.
2) Add a direct submodule and require that it move in lock-step with the rest of IREE's deps (mostly driven by the opaque Google LLVM integrate).
3) Add a forked submodule and hope for it to move in lock-step but allow patches in the fork as needed.

Isolating this dependency (and the others) can help if compatibility becomes difficult to maintain. For example, we could put deps in [experimental](https://github.com/google/iree/tree/main/experimental) or behind a CMake option so they are easy to disable temporarily.

Reply:
Generally all the projects we depend on that depend on LLVM are kept at the bleeding edge because they're being actively developed alongside MLIR. We have fork repositories of MHLO, TF, and LLVM, so that we can add local patches if we need to. Unclear if that would be necessary here. Generally we would wait for our dependencies to bump. If we're just copying a single file, then I think it makes sense to copy it into third_party. But if we're copying a significant chunk of the repo, as appears to be the case here, I would strongly prefer a submodule. One cost of a submodule is that when it's checked out, you're checking out the entire repo (not just the files you need), but an advantage is that the only thing stored in the git history is a commit hash. I would therefore say that if something is temporary that's even *more* reason to use a submodule rather than copying files.